csrc/scheduler/normalization_inner.cpp

```diff
@@ -252,6 +261,386 @@ int64_t getMaxPersistentBatch(
+  const int64_t min_threads_per_block = 4l * threads_per_warp;
+
+  // set the min persistent buffer size to avoid requesting
+  // a block size larger than device limit
+  const int64_t batches_per_block_inner_reduction_min =
+      ceilDiv(parallel_after_vectorize, max_threads_in_block);
+
+  // set the max persistent batch size to avoid low occupancy
+  // don't increase persistent batch size until we have enough threads
+  const int64_t pbs_max_1 =
+      ceilDiv(parallel_after_vectorize, min_threads_per_block);
+  // derived the maximum persistent batch size from the target occupancy
+  const int64_t buffer_bytes_per_batch = max_persistent_buffer_size /
+      total_reduction_numel * (int64_t)vectorize_factor;
+  const int64_t target_threads_per_sm =
+      std::min(target_warps_per_sm * threads_per_warp, max_threads_per_sm);
+  const int64_t pbs_max_2 = getMaxPersistentBatch(
+      buffer_bytes_per_batch, target_threads_per_sm, register_overhead);
+  const int64_t batches_per_block_inner_reduction_max = std::max(
+      batches_per_block_inner_reduction_min, std::min(pbs_max_1, pbs_max_2));
+
+  // Compute maximum number of reductions we could do in the same kernel based
+  // on persistent buffer size. Bounded by the wave count for utilization of
+  // SMs.
+  const int64_t max_multi_reduction_factor = std::min(
+      scheduler_utils::safeDiv(
+          scheduler_utils::register_file_size, max_persistent_buffer_size),
+      ceilDiv(total_iteration_numel, device_multiprocessor_count));
+
+  // Generate a heuristic for each possible persistent batch size.
+  // record which persistent batch size has the highest occupancy.
+  int64_t idx_max_occupancy = -1;
+  int64_t current_max_occupancy = -1;
+  std::vector<HeuristicParams> all_heuristics;
+  all_heuristics.reserve(
+      batches_per_block_inner_reduction_max -
+      batches_per_block_inner_reduction_min + 1);
+  for (int64_t pbs = batches_per_block_inner_reduction_min;
+       pbs <= batches_per_block_inner_reduction_max;
+       pbs++) {
+    all_heuristics.push_back(getHeuristicParamsGivenPerisisentBatchSize(
+        parallel_after_vectorize,
+        total_iteration_numel,
+        max_multi_reduction_factor,
+        min_threads_per_block,
+        buffer_bytes_per_batch,
+        target_warps_per_sm,
+        register_overhead,
+        pbs));
+    if (all_heuristics.back().occupancy > current_max_occupancy) {
+      current_max_occupancy = all_heuristics.back().occupancy;
+      idx_max_occupancy = (int64_t)all_heuristics.size() - 1;
+    }
+  }
+
+  // Sort the heuristics and select the best one.
+  // If no persistent batch size can achieve the target occupancy, and
+  HeuristicParams best_heuristic;
+  if (current_max_occupancy < target_warps_per_sm) {
+    best_heuristic = all_heuristics.at(idx_max_occupancy);
+  } else {
+    // Prioritize occupancy for all fusions except softmax.
+    // Not sure why softmax has regression (about 10%) if prioritize occupancy.
```

Comment:
I know you are already on top of this. Let's either:
1. figure out why before we merge the PR, or
2. open an issue and follow up on that in a later PR.

Reply:
@naoyam softmax is also causing trouble in this PR.

Reply:
What do you mean?

Reply:
Not sure why softmax has regression (about 10%) if prioritize occupancy. 
All other tested fusions show improvement using the following method to select the best persistent batch sizes.
```
// (1) If the maximum occupancy is less than the target occupancy, use the batch
//     leads to the largest occupancy.
// (2) Otherwise, sort the heuristics to prioritize occupancy.
//     (a) Prefer occupancy larger than target.
//     (b) Prefer divisible by persistent batch size.
//     (c) Prefer large register overhead.
//     (d) Prefer large occupancy.
//     (e) Tiebreaker, use large persistent batch size.
// (3) For softmax, a special rule is used to reduce inter-thread communication
//     (a) Prefer bdimx_val equals power of 2
//     (b) Prefer occupancy larger than target.
//     (c) Tiebreaker, use small bdimx
```

Reply:
some explanations: https://docs.google.com/document/d/1oKG3Tb_abbNnqV7EUMU5GveUUwjK6DvON0jrudKTPp8/edit#heading=h.rp1ug14ktzrc

Reply:
I added some explanations on why softmax is so special and creasted an issue to track: https://github.com/NVIDIA/Fuser/issues/1876